programs/bpf/rust/instruction_introspection/src/lib.rs

```diff
@@ -0,0 +1,46 @@
+//! @brief Example Rust-based BPF program that exercises instruction introspection
+
+extern crate solana_sdk;
+use solana_sdk::{
+    account_info::next_account_info, account_info::AccountInfo, entrypoint,
+    entrypoint::ProgramResult, info, program_error::ProgramError, pubkey::Pubkey,
+};
+
+entrypoint!(process_instruction);
+fn process_instruction(
+    _program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    instruction_data: &[u8],
+) -> ProgramResult {
+    if instruction_data.is_empty() {
+        return Err(ProgramError::InvalidAccountData);
+    }
+
+    let secp_instruction_index = instruction_data[0];
+    let account_info_iter = &mut accounts.iter();
+    let instruction_accounts = next_account_info(account_info_iter)?;
+    assert_eq!(
+        *instruction_accounts.key,
+        solana_sdk::sysvar::instructions::id()
+    );
+    let data_len = instruction_accounts.try_borrow_data()?.len();
+    if data_len < 2 {
+        return Err(ProgramError::InvalidAccountData);
+    }
+
+    let instruction = solana_sdk::message::Message::deserialize_instruction(
```

Comment:
Could provide. helper in `sysvar::instructions` to do this (and internally call message, that way the interface to to the instructions sysvar looks encapsulated.  Also, it's a little strange to use `Message` here since there is no indication to the caller that a message is involed.

Reply:
good idea